File: src/main/java/linq/AbstractIterator.java
package linq;

import java.util.List;
import java.util.function.Function;
import java.util.function.Predicate;

public abstract class AbstractIterator<TSource> extends Enumerable<TSource> implements Enumerator<TSource> {
    private final long threadId = Thread.currentThread().threadId();
    protected int state = 0;
    protected TSource current = null;

    @Override
    public TSource current() {
        return current;
    }

    protected abstract AbstractIterator<TSource> getClone();

    @Override
    public void close() {
        current = null;
        state = -1;
    }

    @Override
    public Enumerator<TSource> enumerator() {
        var enumerator = state == 0 && threadId == Thread.currentThread().threadId() ? this : getClone();
        enumerator.state = 1;
        return enumerator;
    }

    public abstract boolean moveNext();


    public <TResult> Enumerable<TResult> select(Function<TSource, TResult> selector) {
        return new Select.EnumerableSelectIterator<>(this, selector);
    }

    public Enumerable<TSource> where(Predicate<TSource> predicate) {
        return new Where.EnumerableWhereIterator<>(this, predicate);
    }

    //    @Override
    //    public void reset() {
    //        throw new UnsupportedOperationException();
    //    }

    public abstract List<TSource> toList();

    public abstract int getCount();

    public AbstractIterator<TSource> skip(int count) {
        return new SkipTake.EnumerableSkipTakeIterator<>(this, count, -1);
    }

    public AbstractIterator<TSource> take(int count) {
        return new SkipTake.EnumerableSkipTakeIterator<>(this, 0, count - 1);
    }

    public TSource tryGetFirst(boolean[] found) {
        return First.tryGetFirstNonIterator(this, found);
    }
}

File: src/main/kotlin/linq/AbstractIterator.kt
package linq

abstract class AbstractIterator<TSource> : Enumerable<TSource>(), Enumerator<TSource> {
    private val threadId = Thread.currentThread().threadId()
    protected var state = 0
    private var backingCurrent: TSource? = null

    override var current: TSource
        get() = backingCurrent ?: throw IllegalStateException()
        set(value) {
            backingCurrent = value
        }

    protected abstract fun clone(): AbstractIterator<TSource>

    override fun close() {
        backingCurrent = null
        state = -1
    }

    override fun enumerator(): Enumerator<TSource> {
        val enumerator = if (state == 0 && threadId == Thread.currentThread().threadId()) this else clone()
        enumerator.state = 1
        return enumerator
    }

    abstract override fun moveNext(): Boolean

    open fun <TResult> itSelect(selector: (TSource) -> TResult): Enumerable<TResult> =
        EnumerableSelectIterator(this, selector)

    open fun itWhere(predicate: (TSource) -> Boolean): Enumerable<TSource> =
        EnumerableWhereIterator(this, predicate)

    abstract fun getList(): List<TSource>

    abstract fun itCount(): Int

    open fun itSkip(count: Int): AbstractIterator<TSource>? =
        EnumerableSkipTakeIterator(this, count, -1)


    open fun itTake(count: Int): AbstractIterator<TSource>? {
        return EnumerableSkipTakeIterator(this, 0, count - 1)
    }

    open fun tryGetElementAt(index: Int): TSource? =
        if (index == 0) {
            tryGetFirst()
        } else {
            tryGetElementAtNonIterator(this, index)
        }

    open fun tryGetFirst(): TSource? = tryGetFirstNonIterator(this)

    open fun tryGetLast(): TSource? = tryGetLastNonIterator(this)

    open fun itContains(value: TSource): Boolean = containsIterate(this, value, null)
}